### ISSUE TITLE
Adjust healthcheck for `prometheus_remote_write`

### ISSUE THREAD

**Issue by jszwedko** [timberio/vector#8279]
Broken off of https://github.com/timberio/vector/pull/8269#pullrequestreview-706306727

The `prometheus_remote_write` sink currently makes a GET request to the endpoint for the health check. This fails for Prometheus itself as its healthcheck URL is `/-/healthy`. The [remote write protocol](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit) does not describe a health check mechanism so it is unlikely that there is an HTTP request that would work for all applications that support the remote write protocol. Instead, I think we may want to just check for connectivity.

cc/ @bruceg in case you know something I don't.

**Comment by spencergilbert** [comment]
I don't think we _currently_ do this anywhere, but we could expose more configuration for `healthcheck` and provide a default (`/-/healthy`) and allow the user to change it if whatever target they're using has a different endpoint?

We'd want to check that other targets do have different endpoints, but I could see this being useful in other sinks as well (`loki` with Grafana Cloud not exposing the healthcheck endpoint).

**Comment by bruceg** [comment]
@jszwedko that matches my knowledge as well. The remote write protocol only has one action, and that is to write metrics. We could possibly submit a request with zero events in it, but I have no idea if that wouldn't cause an error due to being empty. We would have to play with that to determine if it is viable.

**Comment by spencergilbert** [comment]
I noted in Slack as well, but there will be differences in the endpoints per remote_write target - we may want to create a `healthcheck.endpoint`, I imagine that could be useful for other sinks as well

**Comment by zamazan4ik** [comment]
@spencergilbert @jszwedko @bruceg from your point of view, is it ok to add a custom `healthcheck.endpoint` support to the `prometheus_remote_write`? I think I can try to implement it. It defintely would be useful for the users.

**Comment by spencergilbert** [comment]
Talking with @bruceg I think we'd support a PR for that, it should also close https://github.com/vectordotdev/vector/issues/13890. I think `healtcheck.endpoint` is a suitable name for the option, and it could either default to `None` to re-use the `endpoint` configuration - or have a default value of that's a reasonable default.

**Comment by spencergilbert** [comment]
Another followup, talking with @jszwedko - perhaps `healthcheck.path` could be better if we're expecting the base of the address to stay the same and just vary the path, as it would save the user repeating most of the `endpoint` option.

**Comment by zamazan4ik** [comment]
> Another followup, talking with @jszwedko - perhaps healthcheck.path could be better if we're expecting the base of the address to stay the same and just vary the path, as it would save the user repeating most of the endpoint option.

That is an interesting question. From the one hand, specifying only a path will help with path duplication and would prevent possible configuration errors like "changed remote uri but forgot to change a corresponding health check uri". From another hand, in real life possible some really tricky configuration . E.g. remote uri and the corresponding healthcheck uri can be located on the different servers behind different reverse proxies. And in this case we will not be able to configure it, if only a path part can be changed for the healthcheck uri. Or even a user want to define their own dedicated health server, which implements some logic on a health calculation.

Since I did not hear before about so custom setups with different healthcheck uri, I guess we can start with a `healthcheck.path` way. Even it is less flexible, it will reduce possible missconfigurations. Later, if will be requests from users about adding more flexibility, we can think about it and add an additional option or refactor somehow an existing one.

**Comment by Prajwalprakash3722** [comment]
This issue is still persisting, is there any alternatives, or hacky Fixes?

**Comment by zamazan4ik** [comment]
> This issue is still persisting, is there any alternatives, or hacky Fixes?

AFAIK, no fixes yet in this field.

**Comment by akuzia** [comment]
This feature would make Vector one of the best tools for pushing metrics for edge/iot and other uses with limited drive space. Prometheus agent is [kinda](https://github.com/prometheus/prometheus/issues/10750) not function properly in that scenario

**Comment by maulerjan** [comment]
So Vector currently can't write metrics to Mimir?

**Comment by robinpecha** [comment]
+1

**Comment by sojjan1337** [comment]
+1 using 0.41.0 chart. 
prometheus remote_write healthcheck fails. Unexpected status: 405 Method not allowed

**Comment by alxmatweew** [comment]
ver = 0.46.1 ... The problem is relevant and does not allow us to consider Vector as a tool for metrics ((( The "endpoint" parameter for "healthcheck" is really needed

**Comment by josh-a1x** [comment]
+1 at ver=0.46.1
We hit this in our development when attempting to use Vector to push metrics to AWS Managed Prometheus. Would appreciate a fix

**Comment by fsvm88** [comment]
I don't know when it was added, but you can set `healthcheck.enabled: false` for the prometheus-remote-write sink, and it'll stop trying to do a check that the protocol doesn't support. I guess worst case, if prometheus is down, you'll get write errors in the logs... ( [doc here](
https://vector.dev/docs/reference/configuration/sinks/prometheus_remote_write/#healthcheck) )

**Comment by josh-a1x** [comment]
> I don't know when it was added, but you can set `healthcheck.enabled: false` for the prometheus-remote-write sink, and it'll stop trying to do a check that the protocol doesn't support. I guess worst case, if prometheus is down, you'll get write errors in the logs... ( [doc here](
> https://vector.dev/docs/reference/configuration/sinks/prometheus_remote_write/#healthcheck) )

Yes we've been using this successfully for a while

**Comment by alxmatweew** [comment]
> Не знаю, когда это было добавлено, но можно настроить `healthcheck.enabled: false`приемник Prometheus-Remote-Write, и он перестанет пытаться выполнять проверки, не поддерживаемые протоколом. Полагаю, в худшем случае, если Prometheus не работает, в журналах будут ошибки записи... ( [документация здесь](https://vector.dev/docs/reference/configuration/sinks/prometheus_remote_write/#healthcheck) ).

The ability to disable the check is a known fact, but the problem is that for a solution on an enterprise environment in our company this is unacceptable, and therefore there is no possibility of using Vector.

**Comment by alxmatweew** [comment]
> I don't know when it was added, but you can set `healthcheck.enabled: false` for the prometheus-remote-write sink, and it'll stop trying to do a check that the protocol doesn't support. I guess worst case, if prometheus is down, you'll get write errors in the logs... ( [doc here](https://vector.dev/docs/reference/configuration/sinks/prometheus_remote_write/#healthcheck) )

The ability to disable the check is a known fact, but the problem is that for a solution on an enterprise environment in our company this is unacceptable, and therefore there is no possibility of using Vector.

**Comment by simonhammes** [comment]
I'll be working on this.

> Since I did not hear before about so custom setups with different healthcheck uri, I guess we can start with a healthcheck.path way. Even if it is less flexible, it will reduce a chance of possible missconfigurations. Later, if will be the requests from users about adding more flexibility, we can think about it and add an additional option or refactor somehow an existing one.

Since there's already `healthcheck.uri` in other sinks (which is just not respected by the `prometheus_remote_write` sink), I'd argue in favor of using the already existing config setting.

> What about hardcoding healthcheck url just for prometheus_remote_write, something like $endpoint"/-/healthy"

This does not sound like a good solution to me since the remote write protocol is [implemented by a few different solutions]( https://prometheus.io/docs/specs/prw/remote_write_spec/#compatible-senders-and-receivers) and this would be a breaking change.

I had a look at two implementations:

- VictoriaMetrics implements `/health` as well as `/-/healthy` for [compatibility with Prometheus](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/7f3d370244d0fff9f120308c352eb622afb2ac98/lib/httpserver/httpserver.go#L419)
- GreptimeDB does not implement `/-/healthy`, as far as I'm aware.
